### PR TITLE
Alfoa/initialize for codes

### LINKED ISSUE #901: Code Interface initialize method
--------
Issue Description
--------
##### What did you expect to see happen?
The code interface should be able to initialize info  at the begin of the step before an actual perturbation is performed.

##### What did you see instead?
No way to do so

##### Do you have a suggested fix for the development team?
Add an optional method for that.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?
__Normal task, improvement of the code__

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #901

##### What are the significant changes in functionality due to this change request?
added a new optional method for Code interffaces 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
__Fedora 27 is disabled and replaced with Fedora 28__
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/CodeInterfaces/RAVEN/RAVENInterface.py [hunk 1]
```diff
@@ -162,21 +162,16 @@ def generateCommand(self,inputFiles,executable,clargs=None,fargs=None, preExec=N
     returnCommand = executeCommand, outputfile
     return returnCommand
 
-  def createNewInput(self,currentInputFiles,oriInputFiles,samplerType,**Kwargs):
+  def initialize(self, runInfo, oriInputFiles):
```
Comment (wangcj05) [inline framework/CodeInterfaces/RAVEN/RAVENInterface.py]: Should we add this method into the CodeInterfaceBaseClass?

File: doc/user_manual/couplingAcode.tex [hunk 2]
```diff
@@ -311,6 +312,20 @@ \subsubsection{Method: \texttt{getInputExtension}}
...
+\label{subsubsec:codeInterfaceinitialize}
+\begin{lstlisting}[language=python]
+def initialize(self, runInfoDict, oriInputFiles)
+\end{lstlisting}
+The \textbf{initialize} function is an optional method. If present, it is called
+by RAVEN code at the begin of each Step (once per step) involving the particular Code Interface. 
+This method is generally indicated to retrieve information from the RunInfo and/or the Input files.
+\\RAVEN is going to call this function passing in the following arguments:
+\begin{itemize}
+  \item \textbf{\texttt{runInfoDict}}, data type = dictionary: dictionary of the info stored in the run info XML block  ;
```
Comment (wangcj05) [inline doc/user_manual/couplingAcode.tex]: empty space before ";"

File: framework/CodeInterfaces/RAVEN/RAVENInterface.py (hunk 1 above)
Comment (alfoa) [reply framework/CodeInterfaces/RAVEN/RAVENInterface.py]: added

File: doc/user_manual/couplingAcode.tex (hunk 2 above)
Comment (alfoa) [reply doc/user_manual/couplingAcode.tex]: removed

File: framework/Models/Code.py [hunk 3]
```diff
@@ -332,6 +332,10 @@ def initialize(self,runInfoDict,inputFiles,initDict=None):
       else:
         self.raiseAMessage('not found pre-executable '+self.executable,'ExceptedError')
 
+    if 'initialize' in dir(self.code):
```
Comment (wangcj05) [inline framework/Models/Code.py]: We can remove the if check, right? Since we have add this method the code interface base class.

Comment (alfoa) [reply framework/Models/Code.py]: we do not strictly require that the CodeInterfaces inherit from the Base Class so I would keep it for now.

Comment (wangcj05) [reply framework/Models/Code.py]: ok, the changes are good to me.

**Comment by alfoa** [comment]
@wangcj05 is this approved?

**Comment by alfoa** [comment]
@wangcj05 can you merge it?

**Pull request merged**